### DISCUSSION TITLE
[RFD - 0002] - Custom User Roles

### DISCUSSION THREAD

**Discussion by AleksandarCole** [semaphoreci/semaphore#68] [RFD]
# [RFD - 0002] - Custom User Roles

| Authors       | State    |
|---------------|----------|
| @darkofabijan <br> @AleksandarCole <br> @VeljkoMaksimovic  <br> @mattrym | `In progress`   |

## What
Semaphore currently employs role-based access control, where users can have different predefined roles at both the organization level and the project level.

Custom User Roles is a proposed feature that will allow organizations to create their own user roles, assign specific permissions to those roles, and apply them at both the organization and project levels.

## Why
While the default roles available on Semaphore suit most users, there have been several use cases and organization types that require custom adjustments to permissions. Organizations with a large number of members, as well as those adhering to security standards like SOC 2, often need more granular control over permissions.

Implementing this feature would provide maximum flexibility, allowing users to fully customize access rights within their Semaphore organization.
___

### Case 1 - Organization secrets access

The default `Member` role in an organization includes access to the organization secrets page. Several organizations prefer restricting this access to Admin or higher roles, adhering to the principle of least privilege. This restriction helps these organizations comply with their security policies more easily.

However, other organizations rely on the member role having this access, as it aligns with their workflow. Changing this default behavior could cause disruptions for them.

### Case 2 - No ssh access for Contributor role in the project

The default Contributor role in an organization allows users to SSH into jobs. Some organizations prefer to restrict this capability to those with the Admin role on a project, enhancing security and control over sensitive operations.

___

## Solutions

**⚠ NOTE**  
The implementation of this project has already been kicked off. 

Semaphore team working on a project can find links to the: Architecture details, wireframe and tasks breakdown [HERE].(https://renderedtext.operately.com/projects/b1ae334e-848c-4700-9b14-c0302df184cf)

Based on the results of the **security review**, these details will not be public. 

However, we're describing the solution and functionality in more detail below. 

### Role Management Page - Summary

With the rollout of [Project roles and Groups](https://semaphoreci.com/blog/announcing-rbac-enhancements-project-roles-and-groups) the backend system has been switched to full RBAC. 

This allows us to create a new page in Organization Settings that would allow users to manage roles within Semaphore. 

Within this page, for both `organization` and `project` roles users should be able to:
- View existing roles
- Create new roles
- Edit user created roles
- Edit default Semaphore roles (`not available in v1`)
- Delete user created roles
- Delete default Semaphore roles (`not available in v1`)

#### Roles main page

On this page user can see all the roles that exist. 

1. Page is located in the organization settings
2. It is a new section in the settings side menu
3. User can create new roles
4. See existing roles including the quick overview of number of permissions assigned
5. View/Edit the existing role - this will open the `Role single page`

![roles-root](https://github.com/semaphoreci/semaphore/assets/61409859/efd9cc15-47bb-4dae-b28f-2c5bd06e72f6)

#### Roles single page

This page is used to create a new role or edit/view existing one. 

User can select:
- Role name
- Role description
- Permissions that the role will have within the role scope (organization or project)
- **Organization roles only:**  (optional) default project role that will be assigned.

https://github.com/semaphoreci/semaphore/assets/61409859/d6b24bc2-0b4b-4f29-a6a7-05c1d61d8c0b

**⚠ Default project role**  

When creating the organization role, user can decide to automatically give access to all projects to the users with such role assigned. For example, by default all users that have `owner` organization role are automatically assigned `admin` role on every project within organization. 

<img width="454" alt="image" src="https://github.com/semaphoreci/semaphore/assets/61409859/3dd200bc-e952-4ff7-b3f4-0819a35b93d6">

### Organization Role Permissions

List of the permissions that can be assigned to the organization role:

- **organization.activity_monitor.view** - _View organization's activity monitor._
- **organization.audit_logs.manage** - _Manage audit log settings for the organization (such as log streams)._
- **organization.audit_logs.view** - _View audit logs._
- **organization.change_owner** - _Change the owner of the organization._
- **organization.contact_support** - _Contact support on behalf of the organization._
- **organization.custom_roles.manage** - _Modify definition of roles within the organization._
- **organization.custom_roles.view** - _View roles within the organization and permissions they carry._
- **organization.dashboards.manage** - _Create new dashboard views._
- **organization.dashboards.view** - _View the existing dashboards within the organization._
- **organization.delete** - _Delete the organization._
- **organization.general_settings.manage** - _Manage general settings of the organization._
- **organization.general_settings.view** - _View general settings for the organization._
- **organization.groups.manage** - _Manage groups within the organization and modify group members._
- **organization.groups.view** - _View user groups within the organization._
- **organization.ip_allow_list.manage** - _Modify the IP allow list for the organization._
- **organization.ip_allow_list.view** - _View the IP allow list._
- **organization.notifications.manage** - _Modify organization notification settings._
- **organization.notifications.view** - _View organization notification settings._
- **organization.okta.manage** - _Modify existing Okta integrations, or create a new one._
- **organization.okta.view** - _View Okta integration settings for the organization._
- **organization.people.invite** - _Invite new people to the organization._
- **organization.people.manage** - _Remove people from the organization, or change their roles within the organization._
- **organization.people.view** - _View list of people within the organization, together with the roles they have._
- **organization.plans_and_billing.manage** - _Modify billing information or subscription plan._
- **organization.plans_and_billing.view** - _Access the billing page._
- **organization.pre_flight_checks.manage** - _Modify pre-flight checks within the organization._
- **organization.pre_flight_checks.view** - _View pre-flight checks within the organization._
- **organization.projects.create** - _Create a new project within the organization._
- **organization.secrets.manage** - _Manage secrets within the organization._
- **organization.secrets.view** - _View secrets within the organization._
- **organization.secrets_policy_settings.manage** - _Manage secrets policy settings within the organization._
- **organization.secrets_policy_settings.view** - _View existing secrets policy settings._
- **organization.self_hosted_agents.manage** - _Manage self-hosted agents within the organization._
- **organization.self_hosted_agents.view** - _View the list of self-hosted agents within the organization._
- **organization.view** - _Access to the organization. This permission is needed to access any page within the organization domain._

### Project Role Permissions

List of the permissions that can be assigned to the project role:

- **project.access.manage** - _Manage who has access to the project._
- **project.access.view** - _View people, groups and bots that have acces to the project._
- **project.artifacts.delete** - _Remove individual artifacts._
- **project.artifacts.modify_settings** - _Modify artifact settings, such as retention policy._
- **project.artifacts.view** - _Access the artifacts on the project, workflow and job level._
- **project.artifacts.view_settings** - _View artifact settings._
- **project.debug** - _Run debug jobs._
- **project.delete** - _Delete the project._
- **project.deployment_targets.manage** - _Manage deployment targets._
- **project.deployment_targets.view** - _View existing deployment targets._
- **project.flaky_tests.manage** - _Create new filters and notifications for flaky tests._
- **project.flaky_tests.view** - _View the flaky tests page._
- **project.general_settings.manage** - _Modify general settings for the project._
- **project.general_settings.view** - _View general settings for the project._
- **project.insights.manage** - _Change how insights for the project are calculated._
- **project.insights.view** - _View insights and analytics for the project._
- **project.job.attach** - _SSH into the running job, or start a debug session._
- **project.job.rerun** - _Manually rerun jobs or entire workflows_
- **project.job.stop** - _Manually stop running jobs or workflows._
- **project.job.view** - _View jobs on the projects workflow._
- **project.notifications.manage** - _Manage project notifications._
- **project.notifications.view** - _View project notifications._
- **project.pre_flight_checks.manage** - _Modify pre-flight checks for the project._
- **project.pre_flight_checks.view** - _View pre-flight checks for the project._
- **project.repository_info.manage** - _Modify git repository information._
- **project.repository_info.view** - _View git repository information._
- **project.scheduler.manage** - _Modify project tasks._
- **project.scheduler.run_manually** - _Trigger manual runs of any existing tasks._
- **project.scheduler.view** - _View tasks within the project._
- **project.secrets.manage** - _Manage project secrets._
- **project.secrets.view** - _View existing secrets related to the project._
- **project.view** - _Access the project. This permission is needed to see any page within the project._
- **project.workflow.manage** - _Change workflow definition (jobs that will run within the workflow)._
- **project.workflow.view** - _View workflows in the project._

**Comment by mattrym** [comment]
It would be good to have an overview on the role assignments for particular users. For starters, I think including a number of users/groups directly assigned to a role would be enough.

Since there are a few ways to acquire a role (e.g. through Github or through role inheritance), there could be many ways to represent that in a clear way, especially for organizations with broader set of roles.

**Reply by AleksandarCole** [reply]
> It would be good to have an overview on the role assignments for particular users. For starters, I think including a number of users/groups directly assigned to a role would be enough.

@mattrym that's a good idea. You mean something like this:
![roles-metadata](https://github.com/semaphoreci/semaphore/assets/61409859/e8598578-df10-4bc8-a01b-0a1a014ac939)

___

> Since there are a few ways to acquire a role (e.g. through Github or through role inheritance), there could be many ways to represent that in a clear way, especially for organizations with broader set of roles.

I'm a bit less clear on this one. Can you please clarify a bit more?

**Reply by mattrym** [reply]
@AleksandarCole That's exactly what I meant!

As for the latter part of my suggestion, what I meant is that it might not be clear if the numbers here reflect all the users, or just those who were granted the role manually. It would be good to have a breakdown for each role displaying how that role was acquired (through Github, through Okta, through role inheritance, and so on) and the numbers. Even better if we can display some avatars (like for Deployment Targets). That doesn't necessarily need to be placed in that view, but it would be good to have it in general.

**Comment by TomFern** [comment]
We're going to track progress of the documentation for this feature in #87 since we are already planning a "Permissions page"
